File: 114-flatten-binary-tree-to-linked-list/flatten-binary-tree-to-linked-list.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    TreeNode* findTail(TreeNode* root){
        if (root->right == NULL) return root;
        return findTail(root->right);
    }

    void flatten(TreeNode* root) {
        if (root == NULL) return;
        TreeNode* pre_left = root->left;
        TreeNode* pre_right = root->right;

        flatten(pre_right);
        if (pre_left){
            flatten(pre_left);
            root->right = pre_left;
            root->left = NULL;
            findTail(pre_left)->right = pre_right;
        }
    }
};

File: 114-flatten-binary-tree-to-linked-list/flatten-binary-tree-to-linked-list.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def flattenRecursively(self, root):
        # flatten and then return tail
        if root == None or (not root.left and not root.right):
            return root
        
        leftTail = self.flattenRecursively(root.left)
        rightTail = self.flattenRecursively(root.right)

        if leftTail:
            leftTail.right = root.right 
            root.right = root.left
            root.left = None

        return rightTail or leftTail
        


        

    def flatten(self, root: Optional[TreeNode]) -> None:
        """
        Do not return anything, modify root in-place instead.
        """
        self.flattenRecursively(root)
        